### PR TITLE
Docs: Ascent Replay Workflow

### PR DESCRIPTION
**Pull request by lge0303**
create a new pull request to replace the #1322 to avoid merge conflicts

### CODE REVIEW

File: Docs/source/visualization/ascent.rst [hunk 1]
```diff
@@ -176,3 +176,175 @@ When the trigger condition is meet, ``cycle() == 300``, the actions in :code:`tr
               db_name: "cinema_out"
 
 You can view the Cinema Database result by opening :code:`cinema_databases/cinema_out/index.html`.
+
+
+Replay
+-------------------------
```
Comment (ax3l) [inline Docs/source/visualization/ascent.rst]: `.rst` files need those aligned, unfortunately:
```suggestion
Replay
------
```

File: Docs/source/visualization/ascent.rst [hunk 2]
```diff
@@ -176,3 +176,175 @@ When the trigger condition is meet, ``cycle() == 300``, the actions in :code:`tr
...
+
+`Replay <https://ascent.readthedocs.io/en/latest/Utilities.html#getting-data-for-replay>`_ is a utility that allows the user to ‘replay’ Conduit Blueprint HDF5 files saved by Ascent or exported by VisIt (starting in version 3.0 beta) back into Ascent. Replay enables the user or developer to pick specific time steps and load them into Ascent
```
Comment (ax3l) [inline Docs/source/visualization/ascent.rst]: Suggestion to explain all terms:
```suggestion
With Ascent/Conduit, one can store the intermediate data files before the rendering step is applied to custom files.
These so-called *Conduit Blueprint HDF5 files can be *replayed*, i.e. rendered without running the simulation again.
VisIt 3.0+ also supports those files.

`Replay <https://ascent.readthedocs.io/en/latest/Utilities.html#getting-data-for-replay>`_ is a utility that allows the user to "replay" a simulation from aforementioned files and rendering them with Ascent.
Replay enables the user or developer to pick specific time steps and load them for Ascent visualization, without running the simulation again.
```

File: Docs/source/visualization/ascent.rst [hunk 3]
```diff
@@ -176,3 +176,175 @@ When the trigger condition is meet, ``cycle() == 300``, the actions in :code:`tr
...
+
+We'll guide you through the procedure of how to use Replay through a WarpX LWFA example: warpx/tools/Ascent/Replay
```
Comment (ax3l) [inline Docs/source/visualization/ascent.rst]:
```suggestion
We'll guide you through the procedure of how to use Replay through a WarpX LWFA example: `Tools/Ascent/Replay/`
```

File: Docs/source/visualization/ascent.rst [hunk 4]
```diff
@@ -176,3 +176,175 @@ When the trigger condition is meet, ``cycle() == 300``, the actions in :code:`tr
...
+
+1. Get Blueprint Files
+-------------------------
```
Comment (ax3l) [inline Docs/source/visualization/ascent.rst]: Sub-sub-section: https://thomas-cokelaer.info/tutorials/sphinx/rest_syntax.html#headings
```suggestion
Get Blueprint Files
^^^^^^^^^^^^^^^^^^^
```

File: Docs/source/visualization/ascent.rst [hunk 5]
```diff
@@ -176,3 +176,175 @@ When the trigger condition is meet, ``cycle() == 300``, the actions in :code:`tr
...
+    -
+      action: action: "add_extracts"
+      extracts:
+        e1:
+          type: "relay"
+          params:
+            path: "my_output_file_name"
+            protocol: "blueprint/mesh/hdf5"
+
+inputs_3d is LWFA input deck. The data size is 256X256X1024 and the simulation will run for 5000 cycles. WarpX will call ascent every 200 cyces. submit the following job submit script will extract 25 Buleprint HDF5:
```
Comment (ax3l) [inline Docs/source/visualization/ascent.rst]:
```suggestion
`inputs_3d` is LWFA inputs file.
The data size is 256x256x1024 and the simulation will run for 5000 cycles.
WarpX will call ascent every 200 cycles. submit the following job submit script will extract 25 Blueprint HDF5:
```

File: Docs/source/visualization/ascent.rst [hunk 6]
```diff
@@ -176,3 +176,175 @@ When the trigger condition is meet, ``cycle() == 300``, the actions in :code:`tr
...
+   #BSUB -e WarpXe.%J
+
+   module load gcc
+   module load cuda
+
+   export OMP_NUM_THREADS=1
+   jsrun -r 6 -a 1 -g 1 -c 7 -l GPU-CPU -d packed -b rs --smpiargs="-gpu" ./warpx inputs_3d
+
+2. example Actions
+-------------------------
```
Comment (ax3l) [inline Docs/source/visualization/ascent.rst]:
```suggestion
Example Actions
^^^^^^^^^^^^^^^
```

File: Docs/source/visualization/ascent.rst [hunk 7]
```diff
@@ -176,3 +176,175 @@ When the trigger condition is meet, ``cycle() == 300``, the actions in :code:`tr
...
+
+Images of Ex contour with particles could be obtained by replaying on the follow action:
```
Comment (ax3l) [inline Docs/source/visualization/ascent.rst]:
```suggestion
A visualization of the electric field component `Ex` with a contour plot and with added particles can be obtained by replaying with the following Ascent Action:
```

File: Docs/source/visualization/ascent.rst [hunk 8]
```diff
@@ -176,3 +176,175 @@ When the trigger condition is meet, ``cycle() == 300``, the actions in :code:`tr
...
+            image1:
+              bg_color: [1.0, 1.0, 1.0]
+              fg_color: [0.0, 0.0, 0.0]
+              image_prefix: "test%06d"
+              camera:
+                azimuth: 20
+                elevation: 30
+                zoom: 2.5
+
+There are Ascent Actions Examples `Ascent Actions Examples <https://ascent.readthedocs.io/en/latest/Actions/Examples.html#yaml-examples>`_ for you to play.
```
Comment (ax3l) [inline Docs/source/visualization/ascent.rst]:
```suggestion
There are more `Ascent Actions examples available <https://ascent.readthedocs.io/en/latest/Actions/Examples.html#yaml-examples>`_ for you to play.
```

File: Docs/source/visualization/ascent.rst [hunk 9]
```diff
@@ -176,3 +176,175 @@ When the trigger condition is meet, ``cycle() == 300``, the actions in :code:`tr
...
+
+3. Run Replay
+-------------------------
```
Comment (ax3l) [inline Docs/source/visualization/ascent.rst]:
```suggestion
Run Replay
^^^^^^^^^^
```

File: Docs/source/visualization/ascent.rst [hunk 10]
```diff
@@ -176,3 +176,175 @@ When the trigger condition is meet, ``cycle() == 300``, the actions in :code:`tr
...
+
+
+Replay executables are created in the utilities/replay directory of the Ascent instaation. There are two version of replay: mpi version replay_mpi and serial version replay_ser. What you use depends on the data set. Here we use replay_mpi as example.
```
Comment (ax3l) [inline Docs/source/visualization/ascent.rst]:
```suggestion
For Ascent Replay, two command line tools are provided in the utilities/replay directory of the Ascent installation.
There are two version of replay: the MPI-parallel version ``replay_mpi`` and a serial version, ``replay_ser``.
Use an MPI-parallel replay with data sets created with MPI-parallel builds of WarpX.
Here we use ``replay_mpi`` as an example.
```

File: Docs/source/visualization/ascent.rst [hunk 11]
```diff
@@ -176,3 +176,175 @@ When the trigger condition is meet, ``cycle() == 300``, the actions in :code:`tr
...
+  test.cycle_000400.root
+  test.cycle_001000.root
+  test.cycle_002000.root
+
+Replay will loop over these files in the order in which they appear in the file.
+
+The following is an image from replay on test.cycle_000400.root with above ascent_action.yaml
+
+
+.. figure:: WarpX_LWFA_400.png
```
Comment (ax3l) [inline Docs/source/visualization/ascent.rst]: I fear this file is too large (nearly 0.5MB) to be commited into WarpX directly. I can show you another location to store it later on.

**Review by ax3l — CHANGES_REQUESTED** [review]
Thank you for documenting this workflow @lge0303. This is great! :sparkles: 

Can we please add the commit from #1332 here as well? It makes sense to review both together.

Some with to structure the docs and examples futher:
- can you please save the input file snippets just in prose? Keeping it here will make its syntax go outdated very soon.
- do we need a specific bsub file here or can we just cross-reference the submission workflow in other parts of the documentation, please?
- please do not check-in the PNG file, as it will make the source-code repo too large. Maybe add it to the pull-request description instead? :) I can show you how to cross-link it in the documentation from that location.

File: Docs/source/visualization/ascent.rst (hunk 1 above)
Comment (lge0303) [reply Docs/source/visualization/ascent.rst]: 1. You want me move all PR#1332 commit files to ./Docs/src/visualization/, and combine PR#1332 to PR#1331?
2. What do you mean "save the input file alongside the regular examples"? where should I put the input file?
3. I'll remove the bsub file
4. please show me how to croo-link the png file.
Thanks!

**Review by ax3l — APPROVED** [review]
I applied all review comments directly to the PR.
Thanks again for the great docs, @lge0303 ! :sparkles:

**Pull request merged**